walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,140 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request == null) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (events == null || events.size() <= 0) {
+      if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return new ArrayList<>();
+      } else {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> attendeesWithOptional = new ArrayList<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    long duration = request.getDuration();
+
+    List<TimeRange> occupiedTimeRange = new ArrayList<>();
+    List<TimeRange> occupiedTimeRangeWithOptional = new ArrayList<>();
+
+    for (Event event: events) {
+      if (hasAttendeesAttending(event.getAttendees(), attendees)) {
+        occupiedTimeRange.add(event.getWhen());
+      }
+
+      if (hasAttendeesAttending(event.getAttendees(), attendeesWithOptional)) {
+        occupiedTimeRangeWithOptional.add(event.getWhen());
+      }
+    }
+
+    Collection<TimeRange> availableMeetings = getAvailableTimeRange(occupiedTimeRangeWithOptional, duration);
+
+    if (!availableMeetings.isEmpty()) {
+      return availableMeetings;
+    } else {
+      // Has mandatory attendees
+      if (!attendees.isEmpty()) {
+        return getAvailableTimeRange(occupiedTimeRange, duration);
+      } else {
+        return new ArrayList<>();
+      }
+    }
+  }
+
+  private boolean hasAttendeesAttending(Collection<String> eventAttendees, Collection<String> requestAttendees) {
+    for (String attendee: requestAttendees) {
+      if (eventAttendees.contains(attendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private Collection<TimeRange> getAvailableTimeRange(List<TimeRange> occupiedTimeRange, long duration) {
+    Collection<TimeRange> availableMeetings = new ArrayList<>();
+
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return availableMeetings;
+    }
+
+    if (occupiedTimeRange == null || occupiedTimeRange.size() <= 0) {
```

Comment:
occupiedTimeRange is not nullable unless explicitly marked as `@Nullable`. if it can be nullable (say request returns nullable list) then you should do the null check there instead of here and only propagate non-nullable objects through your own code unless explicitly marked as `@Nullable`

Reply:
Fixed